src/pudl/transform/vcerare.py

```diff
@@ -283,107 +289,232 @@ def out_vcerare__hourly_available_capacity_factor(
+            ),
+            raw_vcerare__onshore_wind_power_100m=_get_year(
+                raw_vcerare__onshore_wind_power_100m, year
+            ),
+        )
+
+
+def _check_rows(vce) -> AssetCheckResult | None:
     logger.info("Check VCE RARE hourly table is the expected length")
-    if (length := len(asset_df)) != 136437000:
+    (length,) = duckdb.query("SELECT COUNT(*) FROM vce").fetchone()
+    if length != 136437000:
         return AssetCheckResult(
             passed=False,
             description="Table unexpected length",
             metadata={"table_length": length, "expected_length": 136437000},
         )
+    return None
+
+
+def _check_nulls(vce) -> AssetCheckResult | None:
     logger.info("Check there are no NA values in VCE RARE table (except FIPS)")
-    if asset_df[asset_df.columns.difference(["county_id_fips"])].isna().any().any():
-        return AssetCheckResult(
-            passed=False,
-            description="Found NA values when there should be none",
-            metadata={
-                "NA index values": asset_df[asset_df.isna().any(axis="columns")].index
-            },
-        )
+    columns = [c for c in vce.columns if c != "county_id_fips"]
+    for c in columns:
+        nulls = duckdb.query(f"SELECT {c} FROM vce WHERE {c} IS NULL").fetchall()  # noqa: S608
+        if len(nulls) > 0:
+            return AssetCheckResult(
+                passed=False,
+                description=f"Found NA values in column {c}",
+            )
+    return None
+
+
+def _check_pv_capacity_factor_upper_bound(vce) -> AssetCheckResult | None:
     # Make sure the capacity_factor values are below the expected value
     # There are some solar values that are slightly over 1 due to colder
     # than average panel temperatures.
     logger.info("Check capacity factors in VCE RARE table are between 0 and 1.")
-    if asset_df.capacity_factor_solar_pv.max() > 1.02:
+    cap_oob = duckdb.query(
+        "SELECT * FROM vce WHERE capacity_factor_solar_pv > 1.02"
+    ).fetchall()
+    if len(cap_oob) > 0:
         return AssetCheckResult(
             passed=False,
             description="Found PV capacity factor values greater than 1.02",
         )
-    if asset_df.filter(regex=r"capacity_factor.*wind").max(axis=None) > 1.0:
-        return AssetCheckResult(
-            passed=False,
-            description="Found wind capacity factor values greater than 1.0",
-        )
+    return None
+
+
+def _check_wind_capacity_factor_upper_bound(vce) -> AssetCheckResult | None:
+    columns = [c for c in vce.columns if c.endswith("wind")]
+    for c in columns:
+        cap_oob = duckdb.query(f"SELECT {c} FROM vce WHERE {c} > 1.0").fetchall()  # noqa: S608
+        if len(cap_oob) > 0:  # noqa: S608
+            return AssetCheckResult(
+                passed=False,
+                description=f"Found wind capacity factor values greater than 1.0 in column {c}",
+            )
```

Comment:
As above, rather than fast failing, it would be nice to know all the columns that have problems after a single run.

Reply:
Same